library/ssl_tls.c

```diff
@@ -1672,6 +1672,8 @@ int mbedtls_ssl_set_session( mbedtls_ssl_context *ssl, const mbedtls_ssl_session
                                         session->ciphersuite ) );
             return( MBEDTLS_ERR_SSL_BAD_INPUT_DATA );
         }
+        ssl->handshake->ciphersuite_info = ciphersuite_info;
+        ssl->handshake->key_exchange_mode = MBEDTLS_SSL_TLS1_3_KEY_EXCHANGE_MODE_PSK_ALL;
```

Comment:
> it should not be set here.
> `chiphersuite_info` is decided in ClientHello and `key_exchange_mode` is decided in `ServerHello` .
In client hello, we will send a list of cipher suite, so it can't be decided in ClientHello, instead, we decide it in parse_ServerHello. 
Right, `key_exchange_mode` is decided in `ServerHello`.  But as we all know, we should compute_early_transform after client hello, which is before parse_ServerHello, unfortunately, compute_early_transform depend on the key_exchange_mode indirectly.  So we have to set the key_exchange_mode before.
Set_session happened when re-connection, we can set the info here.

Reply:
> > it should not be set here.
> > `chiphersuite_info` is decided in ClientHello and `key_exchange_mode` is decided in `ServerHello` .
> 
> In client hello, we will send a list of cipher suite, so it can't be decided in ClientHello, instead, we decide it in parse_ServerHello. Right, `key_exchange_mode` is decided in `ServerHello`. But as we all know, we should compute_early_transform after client hello, which is before parse_ServerHello, unfortunately, compute_early_transform depend on the key_exchange_mode indirectly. So we have to set the key_exchange_mode before. Set_session happened when re-connection, we can set the info here.

move to pre_shared_key not here